### ISSUE TITLE
Windows FontImageSourceService assumes MSIX packaging

### ISSUE THREAD

**Issue by jeremy-visionaid** [dotnet/maui#15802]
### Description

The Windows FontRegistrar assumes the application is packaged with MSIX (all paths are prefixed with "ms-appx:///"), so font loading fails if/when MSIX packaging is disabled.

https://github.com/dotnet/maui/blob/main/src/Core/src/Fonts/FontRegistrar.Windows.cs#L11

```
13:05:02:928	Microsoft.Maui.FontManager: Error: Error loading font 'ms-appx:///OpenSans-Regular.ttf'.
```

Switching to AddEmbeddedResourceFont also fails unless using MSIX packaging.

### Steps to Reproduce

1. Create a new MAUI App
2. Disable MSIX Packaging in csproj:

```
		<!-- Disable MSIX -->
		<WindowsPackageType>None</WindowsPackageType>
		<WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>
		<Platforms>x64</Platforms>
		<Platform>x64</Platform>
		<RuntimeIdentifiers>win10-x64</RuntimeIdentifiers>
		<RuntimeIdentifier>win10-x64</RuntimeIdentifier>
		<WindowsAppSdkDeploymentManagerInitialize>false</WindowsAppSdkDeploymentManagerInitialize>
```

Run the app and see error like:
```
13:05:02:928	Microsoft.Maui.FontManager: Error: Error loading font 'ms-appx:///OpenSans-Regular.ttf'.
```

### Link to public reproduction project repository

https://github.com/molesmoke/MauiFontImageSourceTest

### Version with bug

7.0.86

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

10.0.19041.0

### Did you find any workaround?

Could implement a custom registrar service, but haven't tried

### Relevant log output

```shell
13:05:02:928	Microsoft.Maui.FontManager: Error: Error loading font 'ms-appx:///OpenSans-Regular.ttf'.

Microsoft.Maui.FontImageSourceService: Warning: Unable to generate font image ''.

System.ArgumentException: Value does not fall within the expected range.
   at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|20_0(Int32 hr)
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr)
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout._ICanvasTextLayoutFactory.Create(ICanvasResourceCreator resourceCreator, String textString, CanvasTextFormat textFormat, Single requestedWidth, Single requestedHeight)
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout.<>c__DisplayClass11_0.<.ctor>b__0()
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout..ctor(ICanvasResourceCreator resourceCreator, String textString, CanvasTextFormat textFormat, Single requestedWidth, Single requestedHeight)
   at Microsoft.Maui.FontImageSourceService.RenderImageSource(IFontImageSource imageSource, Single scale)
   at Microsoft.Maui.FontImageSourceService.GetImageSourceAsync(IFontImageSource imageSource, Single scale, CancellationToken cancellationToken)
```

> VS bug [#1923183](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1923183)

**Comment by jeremy-visionaid** [comment]
I wonder if as a quick hack, just adding the following to LoadNativeAppFont in FontRegistrar.Windows.cs might be sufficient...

```
            if (File.Exists(filename))
                return filename;
```

**Comment by drasticactions** [comment]
I assume this would go in line with https://github.com/dotnet/maui/issues/10564, @mattleibow?

**Comment by jeremy-visionaid** [comment]
@drasticactions Thanks for pointing that out. This one might be a dupe of #9104

**Comment by jeremy-visionaid** [comment]
I tried a local MAUI build with the above modification. While it addresses the initial font loading error, trying to use it in FontImageSource fails with:

```
System.ArgumentException: Value does not fall within the expected range.
   at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|20_0(Int32 hr)
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr)
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout._ICanvasTextLayoutFactory.Create(ICanvasResourceCreator resourceCreator, String textString, CanvasTextFormat textFormat, Single requestedWidth, Single requestedHeight)
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout.<>c__DisplayClass11_0.<.ctor>b__0()
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout..ctor(ICanvasResourceCreator resourceCreator, String textString, CanvasTextFormat textFormat, Single requestedWidth, Single requestedHeight)
   at Microsoft.Maui.FontImageSourceService.RenderImageSource(IFontImageSource imageSource, Single scale)
   at Microsoft.Maui.FontImageSourceService.GetImageSourceAsync(IFontImageSource imageSource, Single scale, CancellationToken cancellationToken)
Exception thrown: 'System.ArgumentException' in Microsoft.Maui.dll
Exception thrown: 'System.ArgumentException' in WinRT.Runtime.dll
Microsoft.Maui.FontImageSourceService: Warning: Unable to generate font image ''.
```

So some further modifications would also be required.

**Comment by jeremy-visionaid** [comment]
Now I understand what's going on a bit more, I believe this is actually a dupe of #9104. i.e. The ms-appx URI is intentional and the failure was also due to an issue in WinUI3/Win2D. Nonetheless, I still observe the issue on a local build of MAUI release/7.0.2xxsr7 even though all the PRs have been merged there. @mattleibow perhaps this should be closed in favour of reopening #9104?

**Comment by jeremy-visionaid** [comment]
I spoke to soon. The FontManager does manage to load the font's OK, which is what #9104 is about, but FontImageSource is still broken, which my repro project was also testing. So, even though the initial error is resolved on release/7.0.2xxsr7, #9104 is not sufficient to resolve all the problems constructing CanvasTextLayouts:

```
Microsoft.Maui.FontImageSourceService: Warning: Unable to generate font image ''.

System.ArgumentException: Value does not fall within the expected range.
   at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|20_0(Int32 hr)
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr)
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout._ICanvasTextLayoutFactory.Create(ICanvasResourceCreator resourceCreator, String textString, CanvasTextFormat textFormat, Single requestedWidth, Single requestedHeight)
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout.<>c__DisplayClass11_0.<.ctor>b__0()
   at Microsoft.Graphics.Canvas.Text.CanvasTextLayout..ctor(ICanvasResourceCreator resourceCreator, String textString, CanvasTextFormat textFormat, Single requestedWidth, Single requestedHeight)
   at Microsoft.Maui.FontImageSourceService.RenderImageSource(IFontImageSource imageSource, Single scale)
   at Microsoft.Maui.FontImageSourceService.GetImageSourceAsync(IFontImageSource imageSource, Single scale, CancellationToken cancellationToken)
```

**Comment by mattleibow** [comment]
The URI specified in the CanvasTextFormat's FontFamily has an invalid scheme; the scheme may be omitted, or must be one of ms-appx:// or ms-appdata://.

**Comment by mattleibow** [comment]
I had a look and I can't get the FontImageSource implementation to load fonts by file or in unpackaged apps (which load by file). The exception says:

> The URI specified in the CanvasTextFormat's FontFamily has an invalid scheme; the scheme may be omitted, or must be one of ms-appx:// or ms-appdata://.

I will open a bug in Win2D

**Comment by mattleibow** [comment]
Opened this one and will follow up with that team: https://github.com/microsoft/Win2D/issues/941

**Comment by jingliancui** [comment]
met same issue, my product heavy depend on fontawesome.

**Comment by jeremy-visionaid** [comment]
@GuidoNeele I had to use SVG based icons instead

**Comment by BurkusCat** [comment]
> Opened this one and will follow up with that team: [microsoft/Win2D#941](https://github.com/microsoft/Win2D/issues/941)

It doesn't look like there is a huge amount of activity on that repository (and there are bugs affecting WinUI open from 2017). Is there anything that can be done to prioritise [that issue](https://github.com/microsoft/Win2D/issues/941)?

I wouldn't normally ask this but I see this issue is marked as a "p/1" and it feels like it might be a few .NET versions away from being resolved.

**Comment by jeremy-visionaid** [comment]
@BurkusCat I'm not sure how much you can tell about what's going on via GitHub. Seems like most of the work for WinUI & Win2D take part behind closed doors. You can see the work for a similar issue was tracked via a different system: https://github.com/microsoft/Win2D/issues/891

I was watching WinUI 3 for a year waiting for the release blockers to be fixed before I could target an app against it, and one day with no warning around v1.4 the issues were closed and it was finally "good enough". Though both WinUI and MAUI are improving, they're still pretty niggly.

**Comment by jeremy-visionaid** [comment]
Maybe @Redth could give this issue a poke, since he likely knows the folk involved in the related fix?

**Comment by AndreasReitberger** [comment]
Still facing this issue with latest Net8 SR... :( Is there any schedule for this issue?

**Comment by mattleibow** [comment]
We are currently blocked by Win2D/Windows: https://github.com/microsoft/Win2D/issues/941

**Comment by AndreasReitberger** [comment]
Thanks for the update. Sadly not even a reaction on your issue since 3 months 🫠

**Comment by jeremy-visionaid** [comment]
@AndreasReitberger @BurkusCat @GuidoNeele @AathifMahir @mobynet1 maybe you guys wanna make some noise on the upstream ticket https://github.com/microsoft/Win2D/issues/941 to see if that helps with prioritization? It's not really MAUI's fault...

**Comment by BurkusCat** [comment]
Thank you for the update and the comms! I really appreciate its a tricky one to coordinate since it has a dependency but its great to hear the plan for it :)

**Comment by mattleibow** [comment]
Reopening this just until the new Win2D is released and updated. But, SR3 will be ready for it!

**Comment by AndreasReitberger** [comment]
Thank you so much for your help on that 🙂

**Comment by BurkusCat** [comment]
Hey @mattleibow , I thought I'd give the changes a try early. Should the below changes work?

1. I've updated to the latest .NET MAUI version:
`<MauiVersion>8.0.10</MauiVersion>`

2. Updated Win2D for Windows builds
``` xml
  <!-- Windows specifics-->
  <ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">
    <PackageReference Include="Microsoft.Graphics.Win2D" Version="1.2.0" />
  </ItemGroup>
```

| Before above changes:  | After above changes:  | 
|---|---|
| ![image](https://github.com/dotnet/maui/assets/7528322/d6338d63-139a-4daa-a7de-464dfa151422)  |  ![image](https://github.com/dotnet/maui/assets/7528322/a79a6174-3252-440f-acc9-58e8369e8a5f) | 

With CLR exceptions turned on, I get this error:
```
System.ArgumentException: 'The parameter is incorrect.

The URI specified in the CanvasTextFormat's FontFamily has an invalid scheme; the scheme may be omitted, or must be one of ms-appx:// or ms-appdata://.'
```

**Comment by AndreasReitberger** [comment]
Facing the same issue as @BurkusCat 
I updated to `8.0.10` and set WinUI to `1.2.0`.

Fonts still are missing and icons are shown as `[ ]`

![image](https://github.com/dotnet/maui/assets/2952443/244660d3-b21a-4d83-94ae-ebef4fabb537)

@mattleibow any chance for a feedback from you? Should this already work or do we need to wait for SR4?
Thank you :)

**Comment by AndreasReitberger** [comment]
I opened a new issue here: https://github.com/dotnet/maui/issues/21333

**Comment by jeremy-visionaid** [comment]
Just to note here as well as on #21333, that this works OK for me with MAUI 8.0.14 and WinUI 1.2.0.

**Comment by pouchbunny** [comment]
I was having a problem all of a sudden today even though my app was working just fine, but then the font doesn't fall within the expected etc. One of the fonts was mistyped as .ttv not .ttf, but then it did it for another font and I'm like wtf, the OpenSans ones were just fine but the only difference between this other font and the three others I have was that there were spaces. I removed them and the exception went away DESPITE, again, working until after I updated to the most recent build 8.0.1 and continued seeing it after upgrading to 9.0.0.

**Comment by jingliancui** [comment]
> I was having a problem all of a sudden today even though my app was working just fine, but then the font doesn't fall within the expected etc. One of the fonts was mistyped as .ttv not .ttf, but then it did it for another font and I'm like wtf, the OpenSans ones were just fine but the only difference between this other font and the three others I have was that there were spaces. I removed them and the exception went away DESPITE, again, working until after I updated to the most recent build 8.0.1 and continued seeing it after upgrading to 9.0.0.

I think you would better to raise a new issue.
